posts/2021-05-14-paying-homage-to-the-compall-michael-representation.adoc

```diff
@@ -0,0 +1,466 @@
+rocket.launch();
+
+// Will think it has travelled 100km rather than 160km
+rocket.distanceTravelled
+----
+
+Similarly, as these are just values of type `Double`, with just these plain type aliases we are unable to enforce certain constraints such as https://en.wikipedia.org/wiki/Distance#Distance_versus_directed_distance_and_displacement["distance cannot be negative"]. Ideally we want to https://blog.janestreet.com/effective-ml-revisited/[make this kind of illegal state unrepresentable].
+
+== Case class wrappers
+The simplest approach to achieve type-safety and catch the bug of mixing up miles and kilometres at compile-time would be to create distinct new "wrapper" types. In the following code snippet, we introduce two case classes `Kilometers` and `Miles` that each "wrap" their real underlying `Double` value. And because these are distinct types we are not able to just simply operate on miles when kilometres is the expected unit of measure. The case classes provide a way to create values of `Kilometers` and `Miles` and we provide two helper functions `Units.add` and `Units.toKilometers` for adding two values of `Kilometers` and for converting from `Miles` to `Kilometers`.
+
+[source,scala]
+----
+object Units {
+  final case class Kilometers(value: Double)
+  final case class Miles(value: Double)
+
+  def add(km1: Kilometers, km2: Kilometers): Kilometers = Kilometers(km1.value + km2.value)
+  def toKilometers(miles: Miles): Kilometers = Kilometers(miles.value * 1.6)
+}
+
+import Units._
+
+class Booster() {
+  def provideLaunchBoost(): Miles = Miles(100)
+}
+
+class Rocket(booster: Booster) {
+  private var distance: Kilometers = Kilometers(0)
+
+  def launch(): Unit = {
+    // Kilometers and Miles are different types. So compiler prevents the previous bug
+    val launchBoost: Kilometers = toKilometers(booster.provideLaunchBoost())
+    distance = add(distance, launchBoost)
+  }
+
+  def distanceTravelled: Kilometers = distance
+}
+
+val rocket: Rocket = new Rocket(new Booster())
+rocket.launch();
+
+// Will represent the correct distance travelled
+rocket.distanceTravelled
+----
+
+Furthermore, if we wanted to enforce the constraints about distance only being positive, we could take the "smart constructor" approach to restrict how we obtain values of type `Kilometers` and `Miles`. There are quite a number of ways to achieve this in Scala. For a good explanation of the various ways to implement smart constructors I invite you to check out https://tuleism.github.io/blog/2020/scala-smart-constructors/[this recent blog post]. As I am using Scala 2.13.5 to compile these examples, I will take advantage of the newer cleaner approach of using a `private[foo]` access modifier that is possible with the `-Xsource:3` compiler option. (If you are on a version of Scala pre-2.13.2, then I would personally recommend the approach of using `sealed abstract case class`, which is also explained in the linked blog post).
```

Comment:
I like all the references to various articles you have. This one about smart constructors is good.

Reply:
yep, that's a really nice explanation of the topic and different approaches one could take.